pkg/api/resolver.go

```diff
@@ -0,0 +1,358 @@
+	}
+	previous := previousTime.Format(dateFormat)
+
+	// The keywords that indicate a blocker of some sort
+	// TODO - possibly include this as config option somewhere?
+	b := []string{"TestBlocker", "OpsBlocker"}
+
+	// Get the breakdowns for each set
+	all, err := r.dbClient.GetBreakdown(previous, datestamp, components, nil, false, targets)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get breakdown: %v", err)
+	}
+	blockers, err := r.dbClient.GetBreakdown(previous, datestamp, components, b, false, targets)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get breakdown: %v", err)
+	}
+	custCases, err := r.dbClient.GetBreakdown(previous, datestamp, components, nil, true, targets)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get breakdown: %v", err)
+	}
+
+	return &RollupResolver{
+		all:           &BreakdownResolver{all},
+		blockers:      &BreakdownResolver{blockers},
+		customerCases: &BreakdownResolver{custCases},
+	}, nil
+}
+
+// getRollups gets a list of rollups for every day between startDate and endDate
+// Dates with no data will be have a nil rollup
+// Input is assumed to be parsed
+func (r Resolver) getRollups(startDate, endDate time.Time, components, targets []string) ([]*RollupResolver, error) {
+	var rollups []*RollupResolver
+
+	// Loop over days between start and end date (inclusive)
+	for d := startDate; !d.Equal(endDate.AddDate(0, 0, 1)); d = d.AddDate(0, 0, 1) {
+		rollup, err := r.getRollup(d.Format(dateFormat), components, targets)
+		if err != nil {
+			// Errors for missing data are expected
+			// If unexpected error, log and continue with a nil rollup
+			_, isExpected := err.(*apiSafeError)
+			if !isExpected {
+				log.Printf("Error getting rollup for date '%v': %v", d, err)
+			}
+			rollup = nil
+		}
+		rollups = append(rollups, rollup)
+	}
+
+	return rollups, nil
+}
+
+// getRelease is a helper function to get a release rollup for the given release/components
+func (r Resolver) getRelease(name string, components []string) (*ReleaseResolver, error) {
+	// Lookup release info by name
+	var thisRelease options.Release
+	for i := range r.releases {
+		if name == r.releases[i].Name {
+			thisRelease = r.releases[i]
+			break
+		}
+	}
+	// If no release was found matching name, error out
+	if thisRelease.Name == "" {
+		err := fmt.Errorf("resolver: cannot find release with name %s", name)
+		return nil, newAPISafeError(err, "cannot find release with name %s", name)
+	}
+
+	// Parse dates
+
+	// The GA date will default to latest
+	end, err := r.parseDatestamp(thisRelease.Dates.Ga)
+	if err != nil {
+		safe, err := safeError(err, "Unable to parse date")
+		log.Printf("Error parsing date: %v", err)
+		return nil, safe
+	}
+	endDate, err := time.Parse(dateFormat, end)
+	if err != nil {
+		return nil, newAPISafeError(err, "error parsing date %s", end)
+	}
+
+	// The start date will default to 9 weeks (3 sprints) before the end date.
```

Comment:
Consider looking for first time the release was mentioned in the db and using that date.